File: core/engine.py
"""ConsciousnessEngine: central orchestrator that wires all subsystems
together and runs the simulation loop."""

from __future__ import annotations
import numpy as np

from .config import SimConfig
from .interrelations import InterrelationMatrix
from .dynamics import DynamicsEngine
from .environment import Environment
from .analyzer import SystemAnalyzer
from .personality import PersonalitySystem
from .memory import MemorySystem
from .energy import EnergySystem
from .visualization import ConsciousnessVisualizer


class ConsciousnessEngine:
    """Central orchestrator for the consciousness simulation.

    Each tick executes subsystems in order:
      1. Energy — compute available energy and arousal modifiers
      2. Environment — generate stimuli + poll user input
      3. Personality — compute bias forces, register conflict
      4. Interrelation — propagate inter-aspect influences
      5. Dynamics — update weights with momentum, adaptive LR, homeostasis
      6. Memory — record state, compute memory-based influences
      7. Analyzer — detect emergent patterns
      8. Visualization — render frame
    """

    def __init__(self, config: SimConfig | None = None):
        self.config = config or SimConfig()
        self.tick = 0

        # Weight state
        self.weight_dict = self.config.get_initial_weights()
        self._initial_abs_sum = sum(abs(v) for v in self.weight_dict.values())

        # Rolling history
        self.weight_history: list[list[float]] = []
        self.tick_history: list[int] = []

        # Initialize all subsystems
        aspects = self.config.aspects
        initial_weights = self.config.get_initial_weights()

        # Interrelation matrix
        self.interrelation = InterrelationMatrix()
        self.interrelation.build(aspects, self.config.interrelationships)

        # Dynamics engine
        self.dynamics = DynamicsEngine()
        self.dynamics.init_state(aspects, initial_weights, self.config)

        # Environment
        self.environment = Environment(aspects, backend=self.config.input_backend)

        # Analyzer
        self.analyzer = SystemAnalyzer(
            aspects, self.interrelation.matrix, window=50)

        # Personality (optional)
        self.personality: PersonalitySystem | None = None
        if self.config.personality_profile:
            self.personality = PersonalitySystem(
                self.config.personality_profile, aspects)

        # Memory
        self.memory = MemorySystem(
            aspects, stm_size=self.config.memory_stm_size,
            ltm_capacity=self.config.memory_ltm_capacity,
            influence_scale=self.config.memory_influence_scale)

        # Energy
        self.energy = EnergySystem(
            aspects, initial_weights,
            max_energy=self.config.energy_budget,
            attention_slots=self.config.attention_slots,
            circadian_period=self.config.circadian_period)

        # Visualization
        self.viz: ConsciousnessVisualizer | None = None
        if self.config.vis_enabled:
            self.viz = ConsciousnessVisualizer(self.config)
            self.viz.setup()

    def _scale_weights(self, weights: list[float]) -> list[float]:
        current = sum(abs(w) for w in weights)
        if current == 0:
            return weights
        factor = self._initial_abs_sum / current
        return [w * factor for w in weights]

    def step(self) -> dict:
        """Execute one simulation tick. Returns tick summary."""
        aspects = self.config.aspects
        self.tick += 1

        # 1. Energy
        energy_state = self.energy.step(self.weight_dict)
        aspect_scales = energy_state['aspect_scales']

        # 2. Environment: stimuli + user input
        weights_list = [self.weight_dict[a] for a in aspects]
        stimuli = self.environment.generate_stimuli(weights_list, self.tick)
        user_input = self.environment.get_user_input()

        # Scale stimuli by energy/attention
        for i, a in enumerate(aspects):
            stimuli[i] *= aspect_scales.get(a, 1.0)
        stimuli = [s * energy_state['noise_modifier'] for s in stimuli]

        # 3. Personality
        personality_biases = None
        if self.personality:
            self.personality.register_input(self.weight_dict, stimuli)
            stimuli = self.personality.apply_conflict_volatility(stimuli)
            personality_biases = self.personality.compute_biases(self.weight_dict)

        # 4. Interrelation matrix
        rebalanced = self.interrelation.propagate(weights_list)

        # 5. Memory influences (from previous tick's state)
        memory_influences = self.memory.compute_influences(self.weight_dict)

        # 6. Dynamics update
        energy_mods = {a: energy_state['lr_modifier'] for a in aspects}
        updated = self.dynamics.update(
            rebalanced, stimuli, aspects,
            energy_modifiers=energy_mods,
            personality_biases=personality_biases,
            memory_influences=memory_influences)

        # Scale to preserve total magnitude
        updated = self._scale_weights(updated)

        # Apply user input
        self.weight_dict = dict(zip(aspects, updated))
        if user_input['active']:
            self.weight_dict = self.environment.apply_input(
                self.weight_dict, user_input)
            self.analyzer.mark_perturbation()
            self.energy.inject_perturbation(0.3)

        # 7. Memory: record current state
        current_stimuli = stimuli
        self.memory.process_frame(
            [self.weight_dict[a] for a in aspects], current_stimuli)

        # 8. Analyzer
        analysis = self.analyzer.tick_update(self.weight_dict)

        # Update history
        snapshot = [self.weight_dict[a] for a in aspects]
        self.weight_history.append(snapshot)
        self.tick_history.append(self.tick)

        window = self.config.vis_rolling_window
        if len(self.weight_history) > window:
            self.weight_history = self.weight_history[-window:]
            self.tick_history = self.tick_history[-window:]

        return {
            'tick': self.tick,
            'weights': dict(self.weight_dict),
            'analysis': analysis,
            'energy': self.energy.get_state(),
            'user_input': user_input,
            'env_status': self.environment.get_status(),
        }

    def render(self, tick_result: dict):
        """Render one frame via the visualization subsystem."""
        if self.viz is None:
            return
        self.viz.render(
            weight_dict=tick_result['weights'],
            interrelation_matrix=self.interrelation.matrix,
            user_input=tick_result['user_input'],
            analysis=tick_result.get('analysis'),
            energy_state=tick_result.get('energy'),
            env_status=tick_result.get('env_status'),
            iteration=tick_result['tick'])

    def print_initial_state(self):
        print('Initial Weights:')
        print(f"{'Aspect':<26}{'Weight':>8}  {'LR':>6}  {'Mom':>6}  {'Category':<12}")
        for a, w in sorted(self.weight_dict.items(), key=lambda x: x[1], reverse=True):
            lr = self.config.get_learning_rate(a)
            mu = self.config.get_momentum(a)
            cat = self.config.aspect_category(a)
            print(f'{a:<26}{w:>8.4f}  {lr:>6.3f}  {mu:>6.2f}  {cat:<12}')
        print()
        if self.personality:
            print(f'Personality: {self.personality.profile.name}')
            print(f'  {self.personality.profile.description}')
            print()

    def run(self):
        """Main entry point: print state, set up animation, run."""
        from matplotlib.animation import FuncAnimation
        import matplotlib.pyplot as plt

        self.print_initial_state()

        # Seed one tick
        result = self.step()

        if self.viz is None:
            # Headless mode
            print('Running headless (no visualization)...')
            try:
                while True:
                    result = self.step()
                    if self.tick % 100 == 0:
                        summary = self.analyzer.summary()
                        print(f"Tick {self.tick}: phase={summary.get('dominant_phase')} "
                              f"entropy={summary.get('entropy', {}).get('label', '?')} "
                              f"energy={result['energy']['energy_pct']}%")
            except KeyboardInterrupt:
                print(f'\nStopped at tick {self.tick}.')
            return

        def animate(frame):
            result = self.step()
            self.render(result)

        ani = FuncAnimation(self.viz.fig, animate,
                            interval=self.config.vis_interval_ms,
                            cache_frame_data=False)
        plt.show()

File: native/src/engine.rs
//! ConsciousnessEngine — central orchestrator wiring all subsystems together.
//!
//! Mirrors the Python ConsciousnessEngine in core/engine.py exactly in terms
//! of subsystem order and arithmetic.

use crate::analyzer::{Attractor, SystemAnalyzer, TickResults};
use crate::behavior::{compute_behavior, BehaviorResult};
use crate::config::SimConfig;
use crate::dynamics::{DynamicsEngine, UpdateInputs};
use crate::dynamics::SimConfig as DynConfig;
use crate::energy::{EnergySystem, EnergyState};
use crate::environment::Environment;
use crate::interrelations::InterrelationMatrix;
use crate::memory::MemorySystem;
use crate::personality::PersonalitySystem;
use crate::rng::Rng;

// ── TickResult ────────────────────────────────────────────────────────────────

pub struct AttractorSnapshot {
    pub basin_radius: f64,
    pub strength: f64,
    pub drift_rate: f64,
}

pub struct TickResult {
    pub tick: u64,
    pub weights: Vec<f64>,
    pub energy: EnergyState,
    pub analysis: TickResults,
    pub attractors: Vec<AttractorSnapshot>,
    pub env_status: String,
    pub behavior: BehaviorResult,
}

// ── ConsciousnessEngine ───────────────────────────────────────────────────────

pub struct ConsciousnessEngine {
    pub config: SimConfig,
    pub personality: Option<PersonalitySystem>,
    tick: u64,
    weights: Vec<f64>,
    initial_abs_sum: f64,
    interrelation: InterrelationMatrix,
    dynamics: DynamicsEngine,
    environment: Environment,
    analyzer: SystemAnalyzer,
    memory: MemorySystem,
    energy: EnergySystem,
    rng: Rng,
    prev_stress: f64,
    current_behavior: String,
}

impl ConsciousnessEngine {
    pub fn new(config: SimConfig) -> Self {
        let n = config.n;
        let initial_weights = config.initial_weights.clone();
        let initial_abs_sum: f64 = initial_weights.iter().map(|w| w.abs()).sum();

        // Interrelation matrix
        let mut interrelation = InterrelationMatrix::new(n);
        interrelation.build(&config.interrelation_triples);

        // Dynamics engine
        let dyn_cfg = DynConfig {
            homeostasis_rate: config.homeostasis_rate,
            hysteresis_threshold: config.hysteresis_threshold,
            hysteresis_resistance: config.hysteresis_resistance,
        };
        let mut dynamics = DynamicsEngine::new(n);
        dynamics.init(
            &initial_weights,
            &config.learning_rates,
            &config.momentum,
            &dyn_cfg,
        );

        // Environment
        let environment = Environment::new(n);

        // Build flat interrelation data for the analyzer (n×n row-major)
        let interrelation_data: Vec<f64> = interrelation.matrix().to_vec();

        // Analyzer
        let analyzer = SystemAnalyzer::new(
            config.aspects.clone(),
            interrelation_data,
            50,
            0.02,
            0.005,
        );

        // Memory
        let memory = MemorySystem::new(
            config.aspects.clone(),
            config.memory_stm_size,
            config.memory_ltm_capacity,
            config.memory_influence_scale,
            200.0, // decay_half_life in frames
        );

        // Energy
        let energy = EnergySystem::new(
            config.aspects.clone(),
            &initial_weights,
            config.energy_budget,
            config.attention_slots,
            config.circadian_period,
        );

        // Personality (built later via set_personality)
        let personality = None;

        Self {
            tick: 0,
            weights: initial_weights,
            initial_abs_sum,
            interrelation,
            dynamics,
            environment,
            analyzer,
            memory,
            energy,
            personality,
            rng: Rng::new(42),
            prev_stress: 0.0,
            current_behavior: String::new(),
            config,
        }
    }

    /// Run one simulation tick. Returns a `TickResult`.
    pub fn step(&mut self) -> TickResult {
        self.tick += 1;
        let n = self.config.n;

        // 1. Energy ─────────────────────────────────────────────────────────
        let energy_state = self.energy.step(&self.weights);

        // 2. Environment: stimuli ───────────────────────────────────────────
        let mut stimuli = self.environment.generate_stimuli(
            &self.weights,
            self.tick,
            &mut self.rng,
            &self.config.aspect_index,
        );

        // Scale stimuli by per-aspect energy scales and global noise modifier
        for i in 0..n {
            stimuli[i] *= energy_state.aspect_scales[i];
        }
        for s in &mut stimuli {
            *s *= energy_state.noise_modifier;
        }

        // 3. Personality ────────────────────────────────────────────────────
        let mut personality_biases: Option<Vec<f64>> = None;
        if let Some(ps) = &mut self.personality {
            ps.register_input(&self.weights, &stimuli);
            ps.apply_conflict_volatility(&mut stimuli, &mut self.rng);
            let mut biases = vec![0.0f64; n];
            ps.compute_biases(&self.weights, &mut biases);
            personality_biases = Some(biases);
        }

        // 4. Interrelation matrix ───────────────────────────────────────────
        let mut rebalanced = vec![0.0f64; n];
        self.interrelation.propagate(&self.weights, &mut rebalanced);

        // 5. Memory influences ──────────────────────────────────────────────
        let memory_influences = self.memory.compute_influences(&self.weights);

        // 6. Dynamics update ────────────────────────────────────────────────
        let energy_mods: Vec<f64> = vec![energy_state.lr_modifier; n];
        let inputs = UpdateInputs {
            energy_modifiers: Some(&energy_mods),
            personality_biases: personality_biases.as_deref(),
            memory_influences: Some(&memory_influences),
        };
        let mut updated = vec![0.0f64; n];
        self.dynamics.update(&rebalanced, &stimuli, n, inputs, &mut updated);

        // Scale to preserve total magnitude
        let current_abs: f64 = updated.iter().map(|w| w.abs()).sum();
        if current_abs > 1e-12 && self.initial_abs_sum > 1e-12 {
            let factor = self.initial_abs_sum / current_abs;
            for w in &mut updated {
                *w *= factor;
            }
        }
        self.weights = updated;

        // 7. Memory: record frame ───────────────────────────────────────────
        self.memory.process_frame(&self.weights, &stimuli);

        // 8. Analyzer ───────────────────────────────────────────────────────
        let analysis = self.analyzer.tick_update(self.weights.clone());

        // Capture attractor snapshots (last 3)
        let attractors: Vec<AttractorSnapshot> = self.analyzer.attractors()
            .iter()
            .rev()
            .take(3)
            .map(|a: &Attractor| AttractorSnapshot {
                basin_radius: a.basin_radius,
                strength: a.strength,
                drift_rate: a.drift_rate,
            })
            .collect::<Vec<_>>()
            .into_iter()
            .rev()
            .collect();

        // Behavior scoring ──────────────────────────────────────────────────
        let current_stress = energy_state.stress;
        let stress_spike = (current_stress - self.prev_stress).max(0.0);
        let behavior = compute_behavior(
            &self.weights,
            &self.config.aspect_index,
            current_stress,
            stress_spike,
            energy_state.energy_pct,
            &self.current_behavior,
        );
        self.current_behavior = behavior.primary.clone();
        self.prev_stress = current_stress;

        // Environment status string
        let env_status = self.environment.get_status();

        TickResult {
            tick: self.tick,
            weights: self.weights.clone(),
            energy: energy_state,
            analysis,
            attractors,
            env_status,
            behavior,
        }
    }

    /// Inject a user input direction ("positive" or "negative").
    pub fn inject_input(&mut self, direction: &str) {
        self.environment.apply_input(direction, &self.config.aspect_index);
        self.analyzer.mark_perturbation();
        self.energy.inject_perturbation(0.3);
    }

    /// Change the personality profile. Pass `None` to disable.
    pub fn set_personality(&mut self, name: Option<&str>) {
        self.personality = name.and_then(|n| {
            PersonalitySystem::from_profile(n, &self.config.aspect_index, self.config.n)
        });
    }

    /// Reset to initial weights and zero tick counter.
    pub fn reset(&mut self) {
        self.weights = self.config.initial_weights.clone();
        self.tick = 0;
        self.prev_stress = 0.0;
        self.current_behavior = String::new();
        self.dynamics.reset();
    }

    pub fn weights_ref(&self) -> &[f64] { &self.weights }
    pub fn dynamics(&self) -> &crate::dynamics::DynamicsEngine { &self.dynamics }

    /// Randomize weights AND per-aspect learning rates and momentum,
    /// creating a unique personality with different response dynamics.
    pub fn randomize_weights(&mut self, rng: &mut crate::rng::Rng) {
        // Randomize starting weights — these are the initial biases
        for w in self.weights.iter_mut() {
            *w = rng.uniform(-0.8, 0.8);
        }
        // Randomize learning rates (how fast each aspect changes)
        // Range: 0.005 to 0.05 — some aspects will be very reactive, others glacial
        for lr in self.dynamics.lr_mut().iter_mut() {
            *lr = rng.uniform(0.005, 0.05);
        }
        // Randomize momentum (how much inertia each aspect has)
        // Range: 0.4 to 0.96 — some aspects flip easily, others carry trajectory
        for mu in self.dynamics.mu_mut().iter_mut() {
            *mu = rng.uniform(0.4, 0.96);
        }
        self.current_behavior = String::new();
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::config::SimConfig;

    #[test]
    fn engine_runs_100_ticks() {
        let config = SimConfig::default_20();
        let mut engine = ConsciousnessEngine::new(config);
        for _ in 0..100 {
            let result = engine.step();
            assert!(!result.behavior.primary.is_empty());
            assert!(result.energy.energy_pct.is_finite());
        }
    }
}
